File: src/host/Integrity.js
// src/host/Integrity.js
//
// Subresource Integrity verification.
//
// Contract section 3.2: a host MUST verify every fetched resource against the
// digest in the profile before executing or instantiating it, and MUST NOT
// offer an option to skip. The profile and the code it names need not share an
// origin, so an unverified profile is an instruction to execute whatever
// currently sits at a URL.

const ALGORITHMS = Object.freeze({
  sha256: 'SHA-256',
  sha384: 'SHA-384',
  sha512: 'SHA-512'
})

/** Parse `sha384-<base64>` into its parts. Throws on anything else. */
export function parseIntegrity (integrity) {
  if (typeof integrity !== 'string' || integrity.length === 0) {
    throw new Error('no integrity digest; a resource without one cannot be loaded')
  }
  const match = /^(sha256|sha384|sha512)-([A-Za-z0-9+/]+={0,2})$/.exec(integrity.trim())
  if (!match) {
    throw new Error(`malformed integrity digest: ${integrity}. Expected sha384-<base64>.`)
  }
  return { algorithm: match[1], subtleName: ALGORITHMS[match[1]], expected: match[2] }
}

function toBase64 (buffer) {
  const bytes = new Uint8Array(buffer)
  let binary = ''
  for (const b of bytes) binary += String.fromCharCode(b)
  return btoa(binary)
}

/** The digest a byte sequence would need to declare, in SRI form. */
export async function digestOf (bytes, algorithm = 'sha384', { subtle = crypto.subtle } = {}) {
  const subtleName = ALGORITHMS[algorithm]
  if (!subtleName) throw new Error(`unsupported digest algorithm: ${algorithm}`)
  const hash = await subtle.digest(subtleName, bytes)
  return `${algorithm}-${toBase64(hash)}`
}

/**
 * Throws unless the bytes match the declared digest.
 *
 * Throws rather than returning false so that a caller cannot ignore the result
 * by forgetting to check it. There is no verify-and-continue path.
 */
export async function verifyIntegrity (bytes, integrity, { subtle = crypto.subtle } = {}) {
  const { algorithm, subtleName, expected } = parseIntegrity(integrity)
  const hash = await subtle.digest(subtleName, bytes)
  const actual = toBase64(hash)
  if (actual !== expected) {
    throw new Error(
      `integrity mismatch: declared ${algorithm}-${expected}, got ${algorithm}-${actual}`
    )
  }
  return `${algorithm}-${actual}`
}

File: native/jigdaw-adapter/src/Integrity.cpp
// native/jigdaw-adapter/src/Integrity.cpp
#include "jigdaw/Integrity.hpp"

#include <openssl/evp.h>

#include <algorithm>
#include <array>

namespace jigdaw {
namespace {

std::string toBase64(const unsigned char* data, size_t length) {
    static constexpr char table[] =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    std::string out;
    out.reserve(((length + 2) / 3) * 4);
    for (size_t i = 0; i < length; i += 3) {
        const unsigned v = (data[i] << 16) |
                           (i + 1 < length ? data[i + 1] << 8 : 0) |
                           (i + 2 < length ? data[i + 2] : 0);
        out += table[(v >> 18) & 0x3f];
        out += table[(v >> 12) & 0x3f];
        out += i + 1 < length ? table[(v >> 6) & 0x3f] : '=';
        out += i + 2 < length ? table[v & 0x3f] : '=';
    }
    return out;
}

const EVP_MD* digestFor(const std::string& algorithm) {
    if (algorithm == "sha256") return EVP_sha256();
    if (algorithm == "sha384") return EVP_sha384();
    if (algorithm == "sha512") return EVP_sha512();
    return nullptr;
}

}  // namespace

std::string digestOf(const std::vector<uint8_t>& bytes, const std::string& algorithm) {
    const EVP_MD* md = digestFor(algorithm);
    if (md == nullptr) return {};

    std::array<unsigned char, EVP_MAX_MD_SIZE> hash{};
    unsigned length = 0;
    EVP_MD_CTX* ctx = EVP_MD_CTX_new();
    EVP_DigestInit_ex(ctx, md, nullptr);
    EVP_DigestUpdate(ctx, bytes.data(), bytes.size());
    EVP_DigestFinal_ex(ctx, hash.data(), &length);
    EVP_MD_CTX_free(ctx);

    return algorithm + "-" + toBase64(hash.data(), length);
}

std::string verifyIntegrity(const std::vector<uint8_t>& bytes, const std::string& declared) {
    if (declared.empty()) {
        return "no integrity digest declared, and a resource without one cannot be loaded";
    }
    const auto dash = declared.find('-');
    if (dash == std::string::npos) return "malformed integrity digest: " + declared;

    const std::string algorithm = declared.substr(0, dash);
    if (digestFor(algorithm) == nullptr) {
        return "unsupported digest algorithm: " + algorithm + ". Expected sha256, sha384 or sha512.";
    }

    const std::string actual = digestOf(bytes, algorithm);
    if (actual != declared) {
        return "integrity mismatch: declared " + declared + ", got " + actual;
    }
    return {};
}

}  // namespace jigdaw
